### PR TITLE
Support Google Tag Manager Data Layer

### PR DESCRIPTION
**Pull request by brobatr**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | N
| Patch: Bug Fix?          | N
| Minor: New Feature?      | Y
| Major: Breaking Change?  | N
| Tests Added + Pass?      | Yes
| Documentation Provided   | Yes (code comments and or markdown)
| Any Dependency Changes?  |
| License                  | Apache License, Version 2.0

See discussion #2565 

The PR allows renaming of the Data Layer.  Together with the bugfix #2620 (previously merged) this allows a generic data layer to be configured, i.e. remove ACDL, rename, without making a breaking change.  If GTM code is on the page and the Data Layer name is changed to the Google default name of dataLayer, the array will be consumed and extended by GTM thus creating a Google Event-Driven Data Layer.  The GDL push() method is available and very similar to that of the ACDL, so all related logic remains the same as before.

Note that support for GTag, the Google global tag, has not been included despite some earlier discussion. This would have been an experimental feature and a significantly more complex implementation.  This could be a future enhancement.

The screenshot below shows the Core Component Data Layer used be Google Tag Manager, with a Google event ID and normal OOTB GTM gtm.start timing event added as normal by GTM.

core-components-google

### CODE REVIEW

File: content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js [hunk 1]
```diff
@@ -574,7 +575,9 @@
      */
     function onDocumentReady() {
         dataLayerEnabled = document.body.hasAttribute("data-cmp-data-layer-enabled");
-        dataLayer = (dataLayerEnabled) ? window.adobeDataLayer = window.adobeDataLayer || [] : undefined;
+        dataLayerName = document.body.getAttribute("data-cmp-data-layer-name");
+        dataLayerName = dataLayerName || "adobeDataLayer";
+        dataLayer = dataLayerEnabled ? window[dataLayerName] = window[dataLayerName] || [] : undefined;
```
Comment (cbogdan0707) [inline content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js:580]: this line is quite confusing, would it be better to break it in 2 lines?, also undefined is the value already set, would it be better to just be an if statement or if you want it to do it in one line, use the && operator even though some people might consider that confusing as well, or maybe if you want to keep it as a ternary use the null value

File: content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js [hunk 2]
```diff
@@ -574,7 +575,9 @@
      */
     function onDocumentReady() {
         dataLayerEnabled = document.body.hasAttribute("data-cmp-data-layer-enabled");
-        dataLayer = (dataLayerEnabled) ? window.adobeDataLayer = window.adobeDataLayer || [] : undefined;
+        dataLayerName = document.body.getAttribute("data-cmp-data-layer-name");
```
Comment (cbogdan0707) [inline content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js:578]: these 2 lines maybe can become 1, wouldn't be better to use the or operator on the same line?

File: content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js (hunk 1 above)
Comment (brobatr) [reply content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js:580]: thanks @cbogdan0707 , I agree this has become all a bit twisted and thanks for picking it up.  How about this.  Not shorter but clearer and no unnecessary undefined declaration:

If (dataLayerEnabled) {
	dataLayerName = document.body.getAttribute("data-cmp-data-layer-name") || "adobeDataLayer";
	dataLayer = window[dataLayerName];
}

Comment (brobatr) [reply content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js]: hi @cbogdan0707, sorted!

File: content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js (hunk 2 above)
Comment (brobatr) [reply content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/clientlibs/site/js/accordion.js]: hi @cbogdan0707 sorted!

**Review by cbogdan0707 — APPROVED** [review]

**Comment by LSantha** [comment]
@brobatr , could you add verification instructions for this enhancement?

**Comment by brobatr** [comment]
> @brobatr , could you please add verification instructions for this enhancement?

hi @LSantha just to check what you would like.  Do you mean to test that the datalayer renaming works?  Or is it related to testing whether Google Tag Manager will consume the data layer correctly and push(stuff) will work OK?

**Comment by LSantha** [comment]
> > @brobatr , could you please add verification instructions for this enhancement?
> 
> hi @LSantha just to check what you would like. Do you mean to test that the datalayer renaming works? Or is it related to testing whether Google Tag Manager will consume the data layer correctly and push(stuff) will work OK?

@brobatr, in principle both but most importantly the later.

**Comment by brobatr** [comment]
hi @LSantha,

I hope this is quick and painless - it should not take more than 15 mins: 

Go to https://tagmanager.google.com/ and add a new account including a web container

Under admin (top menu) choose import container and import the JSON below to this message (options: existing, overwrite).  This just catches events and has a couple of test variables.

[GTM-NPBVPPFV_workspace2.json](https://github.com/adobe/aem-core-wcm-components/files/14150084/GTM-NPBVPPFV_workspace2.json)

Then Admin->Install Google Tag Manager .  This contains the script snippets needed in a page and where to put them.

Edit the CC Page component head.html add add the header snippet before the closing template tag in CRX DE as described.  You can ignore the noscript one for the body.

In conf enable the data layer, change the name to dataLayer and disable the ACDL as described in the[ CC Datalayer docu included in the PR](https://github.com/adobe/aem-core-wcm-components/blob/9808073e05606a5197f7e29dd359ca89d8ce401b/DATA_LAYER_INTEGRATION.md)

Open a page in AEM for example http://localhost:4502/content/wknd/us/en/about-us.html?wcmmode=disabled

Install Google Tag Assistant Chrome extension: https://chromewebstore.google.com/detail/tag-assistant-companion/jmekfmbnaedfebfnmakmokmlfpblbfdm

Click on the blue Tag Assistant icon in the plugins area in Chrome and create a new domain with the URL of the AEM page you’re testing.  In the Chrome tabs a Google debug tab will open and should show something like this:

Screenshot 2024-02-03 at 06 50 42
Screenshot 2024-02-03 at 06 50 26

Working from the last event and message from the bottom up you can check the build of the data layer contents and see both Adobe and default Google events (gtm.dom etc, both with GTM meta such as unique event IDs.  In the variables tab you can see two custom variables are set (dl_eventinfo and dl_eventinfo_lang as configured in GTM) containing Core Components data

Of course if you would like more or there are any issues please let me know

Rob

**Review by LSantha — APPROVED** [review]

**Comment by LSantha** [comment]
Hello @brobatr , thank you for the verification instructions and for your contribution. I could follow the steps and see the new features in action. It will be available in the upcoming release.

**Pull request merged**